runtime/intern.c

```diff
@@ -506,10 +527,10 @@ static void intern_rec(struct caml_intern_state* s,
       case CODE_SHARED8:
         ofs = read8u(s);
       read_shared:
-        CAMLassert (ofs > 0);
-        CAMLassert (ofs <= s->obj_counter);
+        if (!s->compressed) ofs = s->obj_counter - ofs;
+        CAMLassert (ofs < s->obj_counter);
```

Comment:
This was previously checking for ofs > 0 and would accept an ofs == s->obj_counter as valid.
In the non-compressed mode the ofs == 0 will now be accepted  and ofs == s->obj_counter will now be rejected, is that intended?
(I haven't looked too deeply into how the marshal references here work, but wouldn't an offset of zero potentially result in infinite recursion, making the 'CAMLassert(ofs > 0)' potentially worth keeping?)

Reply:
This is a tricky change but I think it's OK.  Let me demonstrate.

The `ofs` variable contains one of two quantities:
- POS, the absolute position of the shared object in `intern_obj_table`
- DIST, the relative distance of the shared object to the end (the size) of `intern_obj_table`, which I'll write SIZE.

The equation POS + DIST = SIZE holds at all times.

In the old code, `ofs` is always DIST, the assertion checks 0 < DIST <= SIZE, and then we access the table at POS = SIZE - DIST.

In the new code, in non-compressed mode, `ofs` starts as DIST, then is changed to contain POS = SIZE - DIST, then the assertion checks POS < SIZE.  (It is always the case that 0 <= POS because it's an unsigned integer.)  The we access the table at POS.

And in compressed mode, `ofs` starts as POS, and we check POS < SIZE then access the table at POS.

It's easy to see that 0 < DIST <= SIZE is equivalent to 0 <= POS < SIZE when POS = SIZE - DIST.

So we're checking the same thing as before, namely that POS is within bounds for accessing the table of shared objects, but with a different formula.

Reply:
Thanks for the detailed explanation, I haven't spotted that the assert has changed position and is after the conditional that transforms it from relative to absolute.

Reply:
Thanks for the detailed explanation, I haven't spotted that the assert has changed position and is after the conditional that transforms it from relative to absolute.